core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,466 @@
+            # Dict mapping each session ID to the list of state hit events for
+            # that session.
+            state_hit_events_session_id_mapping = {}
+            # Dict mapping each state name of the exploration and the list of
+            # unique session IDs that have entered that state.
+            state_session_ids = {
+                state_name: [] for state_name in versioned_exploration.states
+            }
+            session_ids_set = set()
+            # Compute total_hit_count and first_hit_count for the states.
+            for state_hit_event in state_hit_events_for_this_version:
+                state_name = state_hit_event['state_name']
+                state_stats_mapping[state_name].total_hit_count_v1 += 1
+                if state_hit_event['session_id'] not in state_session_ids[
+                        state_name]:
+                    state_stats_mapping[state_name].first_hit_count_v1 += 1
+                    state_session_ids[state_name].append(state_hit_event[
+                        'session_id'])
+                    session_ids_set.add(state_hit_event['session_id'])
+                if state_hit_event['session_id'] not in (
+                        state_hit_events_session_id_mapping):
+                    state_hit_events_session_id_mapping[state_hit_event[
+                        'session_id']] = []
+                state_hit_events_session_id_mapping[state_hit_event[
+                    'session_id']].append(state_hit_event)
+
+            # Compute num_completions for the states.
+            for session_id in session_ids_set:
+                state_hit_event_dicts_sessioned = (
+                    state_hit_events_session_id_mapping[session_id])
+                # All state hit events in the above list are ordered by their
+                # timestamp. This implies that the last state hit event is the
+                # state at which the learner quit the exploration (This is only
+                # True if the last state hit event is not the terminal state).
+                # So, if the last state is not the terminal state, we increment
+                # num_completions for all states except the last state.
+                # Otherwise, we increment num_completions for all the states.
+                for state_hit_event in state_hit_event_dicts_sessioned[:-1]:
+                    state_stats_mapping[state_hit_event[
+                        'state_name']].num_completions_v1 += 1
+                if session_id in completed_session_ids:
+                    state_stats_mapping[state_hit_event_dicts_sessioned[-1][
+                        'state_name']].num_completions_v1 += 1
+
+            # Compute total_answers_count and useful_feedback_count.
+            for value in state_answer_events_for_this_version:
+                for answer in value['submitted_answer_list']:
+                    state_stats_mapping[value[
```

Comment:
Worth just using len(value['submitted_answer_list']) for this? I think len() is an O(1) operation in Python.

In fact if you only need the counts then why not pass only this info over from the map() stage? Answers are big.